### PR TITLE
Introduce new AntClassLoader.getUrl() method to prevent code duplication

### PR DESCRIPTION
**Pull request by jsoref**
AntClassLoader.getUrl

This is for discussion.

Notes:
* Most commits were created by IntelliJ
* Individual commits are mostly self contained
* In general, most commits should not change program behavior at all

This is generally a subset of https://github.com/jsoref/jenkins/commits/java-cleanup-full
 

### Proposed changelog entries

* `Internal:` Internal Java code cleanup

### Submitter checklist

- [ ] JIRA issue is well described
- [ ] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### CODE REVIEW

File: core/src/main/java/jenkins/util/AntClassLoader.java [hunk 1]
```diff
@@ -919,6 +912,27 @@ public URL getResource(String name) {
...
     }
 
+    /**
+     * Finds a matching file by iterating pathComponents.
+     *
+     * @param pathComponents Path to a folder, split into the individual folder names
+     * @param name File to find
+     * @return Url to found object
+     */
+    @CheckForNull
```
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/util/AntClassLoader.java]: Could we use a collection with explicit type in new API? Vector is deprecated

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
+1 for deduplication, but we should avoid using Vector in the new codebase if possible

File: core/src/main/java/jenkins/util/AntClassLoader.java (hunk 1 above)
Comment (jsoref) [reply core/src/main/java/jenkins/util/AntClassLoader.java]: @oleg-nenashev, having failed miserably w/ my last couple PR attempts, I'm going to try something new. I've created https://github.com/jsoref/jenkins/pull/1 which probably implements your suggestion. If you like it, I'll merge it into this branch (and thus into this PR).

Comment (res0nance) [reply core/src/main/java/jenkins/util/AntClassLoader.java]: I commented on that PR

Comment (oleg-nenashev) [reply core/src/main/java/jenkins/util/AntClassLoader.java]: Would be nice if you request my review there. Might be difficult to get to it soon

Comment (jsoref) [reply core/src/main/java/jenkins/util/AntClassLoader.java]: @oleg-nenashev: it won't let me. You're tagged to it.

It seems like @res0nance is ok with the changes there. I'm unavailable until Thursday, but unless someone else comments on that PR, I'll squash those commits together and merge that commit into this PR (thus there will be two commits in this PR).

**Review by oleg-nenashev — APPROVED** [review]
Looks good, thanks for the fix

**Review by res0nance — APPROVED** [review]
This is ok it doesn't include the ArrayList fixes though

**Comment by jsoref** [comment]
@res0nance: please confirm that the updated PR there is correct and then I'll merge it into this PR.

File: core/src/main/java/jenkins/util/AntClassLoader.java [hunk 2]
```diff
@@ -919,6 +911,26 @@ public URL getResource(String name) {
...
+    protected URL getUrl(ArrayList<File> pathComponents, String name) {
```
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/util/AntClassLoader.java]:
```suggestion
    protected URL getUrl(Iterable<File> pathComponents, String name) {
```

Just to make it more generic

**Review by oleg-nenashev — APPROVED** [review]

**Comment by oleg-nenashev** [comment]
@jsoref WDYT about the suggested change?

**Review by fcojfernandez — APPROVED** [review]

**Review by oleg-nenashev — APPROVED** [review]
I plan to merge it tomorrow if no negative feedback

File: core/src/main/java/jenkins/util/AntWithFindResourceClassLoader.java [hunk 3]
```diff
@@ -7,24 +7,23 @@
...
-import java.util.Enumeration;
-import java.util.Vector;
 
 /**
  * As of 1.8.0, {@link org.apache.tools.ant.AntClassLoader} doesn't implement {@link #findResource(String)}
  * in any meaningful way, which breaks fast lookup. Implement it properly.
  */
 public class AntWithFindResourceClassLoader extends AntClassLoader implements Closeable {
-    private final Vector pathComponents;
+    private final ArrayList<File> pathComponents;
```
Comment (zbynek) [inline core/src/main/java/jenkins/util/AntWithFindResourceClassLoader.java]: This field is only used as a copy of parent's `pathComponents` (see the reflection code in constructor). That means `getUrl` is always called with the same first argument: if you remove that argument and use the field directly in `getUrl`, you can drop the copy of  the field in `AntWithFindResourceClassLoader`. Can be done in another PR though.

**Comment by oleg-nenashev** [comment]
Will merge since https://github.com/jenkinsci/jenkins/pull/4254#discussion_r361814126 is not a blocking comment. But I agree it would be a great follow-up

**Pull request merged**